### ISSUE TITLE
Numpy compatibility issue with libopenblas 0.3.30 package on python 3.9

### ISSUE THREAD

**Issue by ethanglaser** [xianyi/OpenBLAS#5341]
When running the following simple numpy workflow with python 3.9 and conda-forge libopenblas 0.3.30, I observe hangs:
`conda create -n reproducer python=3.9 -c conda-forge -y`
`conda activate reproducer`
`conda install numpy -y`
`python -c "import numpy as np; from numpy.testing import assert_allclose; np.linalg.slogdet(np.ones((200,200)))"`

Note that libopenblas (0.3.30) is installed as a dependency of numpy, and the issue is numpy-version agnostic.

Also note that the second python line `from numpy.testing import assert_allclose` is irrelevant to the others, but importing something from numpy.testing seems to complicate the environment and contribute to this issue.

The dimensions, python version, and libopenblas versions are all specific here - with python 3.10, libopenblas 0.3.29, or dim < 200, the python command executes instantaneously. Any idea where this could be originating from?

**Comment by martin-frbg** [comment]
What is your hardware and operating system ? Offhand, I'm not aware of any change in 0.3.30 that could cause this - but I also do not know if anything changed in the build options, compiler etc of condaforge's openblas, or what changes in python between 3.9 and 3.10 could have any bearing on the problem. (Neither would I know immediately what BLAS or LAPACK functions are called by linalg.slogdet, perhaps take this up with the numpy folks first ?)

**Comment by briochh** [comment]
@ethanglaser do you see this behaviour on python versions > 3.9.? I am trying to work out if I am seeing something similar. We are getting CI timeouts for all python version on Ubuntu that seems to trace down to a np.linalg.inv() call. If I pin to libopenblas<0.3.30 everything runs through fine.

**Comment by martin-frbg** [comment]
@briochh I read him as stating that with 3.10 it runs fine...

**Comment by ethanglaser** [comment]
As for hardware, I have faced this issue on a variety of linux Intel platforms (both RHEL+Ubuntu OS, Sapphire Rapids, Ice Lake, Emerald Rapids) and haven't yet found a linux platform where I was unable to reproduce, though I was not able to reproduce on windows. And yes - only on 3.9 (I have not tried lower but did not face the issue on 3.10-3.13). I can open up an issue with numpy as well if you think that would be useful.

**Comment by martin-frbg** [comment]
Thanks - just trying to ascertain where to try to reproduce this. The apparent dependency on matrix size suggests this could be a multithreading effect, maybe python <3.10 was more prone to lockups (whether caused by third-party libraries or python itself). I have since read that slogdet depends on GETRF, which has not changed between 0.3.29 and 0.3.30 - do you see the hangs with other numpy calls as well, and slogdet happened to be the most simple example, or is this specific to slogdet as far as you know ?

**Comment by briochh** [comment]
> @briochh I read him as stating that with 3.10 it runs fine...

Ok, thanks both. I'll do some digging and not try a piggyback on this issue anymore!

**Comment by ethanglaser** [comment]
> do you see the hangs with other numpy calls as well, and slogdet happened to be the most simple example, or is this specific to slogdet as far as you know ?

I have not seen in any others but I haven't tried to go deeper than slogdet. I originally faced this hang when running a scikit-learn test: https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/linear_model/tests/test_bayes.py#L34 and tracked it down to this function. It was ran as part of a test suite of all scikit-learn tests which I imagine would call many numpy functions, this is the only case where it hanged.

**Comment by martin-frbg** [comment]
Hmm. I cannot reproduce this with current `develop` (which should be equivalent to 0.3.30 in any aspect of thread safety at least) in a VM running Debian 11 (which comes with python 3.9.2, numpy 1.19.5) even after increasing the array dimensions to 10000. What's your exact version number of python 3.9 please, and which version of numpy do you use ?

**Comment by ethanglaser** [comment]
Python 3.9.23, numpy 2.0.2 or 1.26.4

**Comment by martin-frbg** [comment]
reproduced with 1.26.4, fix seems to be to declare the server_lock mutex as volatile static instead of just static. hmm...
